---
title: "Test file"
author: "Magnus Sun"
date: "2023-12-04"
output: html_document
---

---
title: "Differential Expression Analysis"
author: "Group 39 Amy"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(BiocManager, tidyverse, reshape2, pheatmap)

# BiocManager::install("DESeq2")
# BiocManager::install("maftools")

library(DESeq2)
library(tidyverse)
```

## Reading Data
```{r}
count_matrix <- read_csv(file = "RNAseq_PRAD.csv")
mutation_data <- read_tsv(file = "data_mutations.txt")
patient_data <- read_tsv(file = "data_clinical_patient.txt", skip = 4)
```

## Find Patients with Complete Data
Get Patient ID in Mutation Data
```{r}
mutation_data <- mutation_data |> mutate(PATIENT_ID = substr(Tumor_Sample_Barcode, 1, 12))
```

Get Patient ID's from RNA Data
```{r}
RNA_ColNames <- count_matrix |> colnames() |> substr(1, 12)
sub_intersect <- intersect(patient_data$PATIENT_ID, mutation_data$PATIENT_ID)
main_intersect <- intersect(sub_intersect, RNA_ColNames)
head(main_intersect)
```

```{r}
test_matrix <- count_matrix[,!names(count_matrix) %in% RNA_ColNames]
```

#Mutation Analysis

##1 Visualize mutation summaries
Find patients in mutation data that exists in all data sets.
```{r}
full_data_patient_ids <- intersect(patient_data$PATIENT_ID, mutation_data$PATIENT_ID) |>
intersect(RNA_ColNames)
mutation_intersect <- subset(mutation_data, PATIENT_ID %in% full_data_patient_ids)
patient_intersect <- subset(patient_data, PATIENT_ID %in% full_data_patient_ids)
colnames(count_matrix) <- colnames(count_matrix)|> substr(1, 12)
count_matrix$...1 <- count_matrix$...1 |> substr(1,15)

count_matrix <- count_matrix[!duplicated(count_matrix$...1),]
count_matrix_gene <- count_matrix$...1 
count_matrix_intersect <- count_matrix[,full_data_patient_ids]

rownames(count_matrix_intersect) <- count_matrix_gene
```

```{r, results='hide'}
library(maftools)
patient_intersect$Tumor_Sample_Barcode <- patient_intersect$PATIENT_ID
laml = read.maf(maf = mutation_intersect, clinicalData = patient_intersect)
```

```{r}
#Writes maf summary to an output file with basename laml.
write.mafSummary(maf = laml, basename = 'laml')
```

```{r}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
```
```{r}
#oncoplot for top twenty mutated genes.
oncoplot(maf = laml, top = 20)
```

## Filter Silent and Missense and Regenerate MAF Summary
```{r}
mutation_filtered <- subset(mutation_intersect, Variant_Classification != "Silent")
mutation_filtered <- subset(mutation_filtered, Variant_Classification != "Missense_Mutation")
mutation_filtered <- subset(mutation_filtered,IMPACT =="HIGH")

count_matrix_filtered <- count_matrix_intersect[,unique(mutation_filtered$PATIENT_ID)]
rownames(count_matrix_filtered) <- count_matrix_gene
```


```{r, results='hide'}
laml.fil = read.maf(maf = mutation_filtered, clinicalData = patient_intersect)
write.mafSummary(maf = laml.fil, basename = 'laml_fil')
```


```{r}
plotmafSummary(maf = laml.fil, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
oncoplot(maf = laml, top = 20)
```


```{r}
hugo <- as.data.frame(table(mutation_filtered$Hugo_Symbol))
```

```{r}
##Create the onco matrix
cnv_events = unique(mutation_filtered$Variant_Classification)
oncomat = reshape2::dcast(
  data = mutation_filtered,
  formula = Hugo_Symbol ~ PATIENT_ID,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
```


```{r}
rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
oncomat.ordered <- oncomat[order(-hugo$Freq),]
mat <- oncomat.ordered
mat[mat!=""]=1
mat[mat==""]=0
mat <- apply(mat, 2 ,as.numeric)
mat <- as.matrix(mat)
rownames(mat)  <-  row.names(oncomat.ordered)
```

```{r}
reduce.mat <- mat[1:10,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE,
         clustering_method = "ward.D")
```
```{r}
cluster <- as.data.frame(cutree(res$tree_col, k = 2)) 

```


## Classify Patients
```{r}
library(dbplyr)
hugo.ordered <- rownames(reduce.mat)

survdata <- patient_intersect |> filter(PATIENT_ID %in% rownames(cluster))

# survdata <- survdata[,c("PATIENT_ID","OS_MONTHS","DSS_MONTHS","OS_STATUS","DAYS_LAST_FOLLOWUP","PFS_STATUS",)]


survdata$Group <- cluster[,1]
survdata$deceased <- grepl("1",survdata$PFS_STATUS,fixed=TRUE)
survdata$overall_survival <- ifelse(survdata$deceased,
                                    survdata$PFS_MONTHS*30,
                                    survdata$DAYS_LAST_FOLLOWUP)
# survdata$overall_survival <- survdata$OS_MONTHS
```

## Survival Analysis
```{r}
library(survival)
library(TCGAbiolinks)
library(survminer)
library(SummarizedExperiment)
fit <- survfit(Surv(overall_survival,deceased)~Group,data=survdata)
ggsurvplot(fit,data=survdata,pval=T)
```



## Expression Analysis
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
countData <- count_matrix_filtered #just to keep the same variable name in the tutorial
colData <- as.data.frame(survdata$Group)
rownames(colData) <- survdata$PATIENT_ID
colnames(colData) <- c("condition")
sampleDists = dist(t(countData),upper = TRUE)


annot_col = data.frame(colData$condition)
row.names(annot_col) <- rownames(colData)

sampleDistMatrix = as.matrix( sampleDists )
rownames(sampleDistMatrix) = colnames(countData)
colnames(sampleDistMatrix) = colnames(countData)

# pheatmap(sampleDistMatrix,
#          clustering_distance_rows = sampleDists,
#          clustering_distance_cols = sampleDists,
#          cluster_rows=FALSE, show_rownames=TRUE,
#          cluster_cols=TRUE,
#          annotation_col=annot_col)
# 
# pca_res <- prcomp(tcountData), scale. = TRUE)
# score <- pca_res$x
# 
# score = as.data.frame(score)
# score$color <- as.factor(colData$condition)
# 
# 
# ggplot(score, aes(x=PC1, y=PC2,  color=color)) +
#   geom_point(size = 4)
```


```{r}
library(DESeq2)
dds = DESeqDataSetFromMatrix(countData=countData,
                              colData=colData,
                              design=~condition)
dds = DESeq(dds)

res1 <- results(dds)
```

```{r}
res05 <- results(dds,alpha=0.05)
summary(res05)
table(res05$padj<0.05)
res05 <- res05[order(res05$pvalue),]
head(res05,10)

```


```{r}
summary(res1)
table(res1$padj<0.05)
```


```{r}
resSig <- subset(res1,padj<0.05)
resSig <- resSig[order(resSig$pvalue),]
head(resSig,10)
```
upregulated means in group2
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
resSig$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(resSig), 
                    column="SYMBOL",
                    keytype="ENSEMBL",
                    multiVals="first")

resSig$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(resSig), 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")

head(resSig$symbol,10)
```

```{r}
#Select top 10 downregulated and top 10 upregulated genes
genesup <- order(resSig$log2FoldChange,decreasing = TRUE)[1:10]
genesdown <- order(resSig$log2FoldChange,decreasing = F)[1:10]
genesSig
tp20udgenes <- append(genesup,genesdown)
resSig[tp20udgenes,]
```


```{r}
vsd <- vst(dds)
```

```{r}
annot_col = data.frame(colData$condition)
row.names(annot_col) <- rownames(colData)

sampleMatrix <- assay(vsd)[tp20udgenes,]

rownames(sampleMatrix) = rownames(countData[tp20udgenes,])
colnames(sampleMatrix) = colnames(countData)

Deseqres <- pheatmap(sampleMatrix , cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, annotation_col=annot_col,clustering_method = "ward.D")

pheatmap(sampleMatrix , cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, annotation_col=annot_col,clustering_method = "ward.D", show_colnames =F ,labels_row = resSig[tp20udgenes,]$symbol)
```

```{r}
# annot_col = data.frame(colData$condition)
# row.names(annot_col) <- rownames(colData)
# 
# sampleMatrix <- assay(vsd)[tp20udgenes,]
# 
# rownames(sampleMatrix) = rownames(countData[tp20udgenes,])
# colnames(sampleMatrix) = colnames(countData)
# 
# Deseqres <- pheatmap(sampleMatrix , cluster_rows=FALSE, show_rownames=TRUE,
#          cluster_cols=TRUE, annotation_col=annot_col,clustering_method = "ward.D")
# 
# pheatmap(sampleMatrix , cluster_rows=FALSE, show_rownames=TRUE,
#          cluster_cols=TRUE, annotation_col=annot_col,clustering_method = "ward.D", show_colnames =F ,labels_row = resSig[tp20udgenes,]$symbol)
```


```{r}
DEcluster <- as.data.frame(cutree(Deseqres$tree_col, k = 2)) 

table(DEcluster)
table(cluster)
```


## pca plot
```{r}
# pca_res <- prcomp(t(countData), scale. = TRUE)
# score <- pca_res$x
# 
# score = as.data.frame(score)
# score$color <- as.factor(colData$condition)
# 
# 
# ggplot(score, aes(x=PC1, y=PC2,  color=color)) + 
#   geom_point(size = 4)
```

```{r}
library(pathview)
library(gage)
library(gageData)
data(kegg.sets.hs)
data(sigmet.idx.hs)
# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)

```

```{r}
foldchanges = resSig$log2FoldChange
names(foldchanges) = resSig$entrez
head(foldchanges)
```

```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)

## Focus on top 10 upregulated and downregulated pathways
keggrespathwaysup <- rownames(keggres$greater)[1:10]
keggrespathwaysdown <- rownames(keggres$less)[1:10]
# Extract the 8 character long IDs part of each string
keggresidsup = substr(keggrespathwaysup, start=1, stop=8)
keggresidsdown = substr(keggrespathwaysdown, start=1, stop=8)
keggresidsup
keggresidsdown
```
```{r}
pathview(gene.data=foldchanges, pathway.id=keggresidsup, species="hsa")


```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresidsdown, species="hsa")
```

