---
title: "Test file"
author: "Magnus Sun"
date: "2023-12-04"
output: html_document
---

---
title: "Differential Expression Analysis"
author: "Group 39 Amy"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(BiocManager, tidyverse, reshape2, pheatmap)

# BiocManager::install("DESeq2")
# BiocManager::install("maftools")

library(DESeq2)
library(tidyverse)
```

## Reading Data
```{r}
count_matrix <- read_csv(file = "RNAseq_PRAD.csv")
mutation_data <- read_tsv(file = "data_mutations.txt")
patient_data <- read_tsv(file = "data_clinical_patient.txt", skip = 4)
```

## Find Patients with Complete Data
Get Patient ID in Mutation Data
```{r}
mutation_data <- mutation_data |> mutate(PATIENT_ID = substr(Tumor_Sample_Barcode, 1, 12))
```

Get Patient ID's from RNA Data
```{r}
RNA_ColNames <- count_matrix |> colnames() |> substr(1, 12)
sub_intersect <- intersect(patient_data$PATIENT_ID, mutation_data$PATIENT_ID)
main_intersect <- intersect(sub_intersect, RNA_ColNames)
head(main_intersect)
rm(RNA_ColNames, sub_intersect)
```
```{r}
patient_data <- patient_data |>
  filter(PATIENT_ID %in% main_intersect);
mutation_data <- mutation_data |>
  filter(PATIENT_ID %in% main_intersect);

ensembl_ids <- count_matrix |> select(...1) |> pull()
keep_cols <- count_matrix |> colnames()
keep_cols <- which(substr(keep_cols, 1, 12) %in% main_intersect)
count_matrix <- count_matrix[, keep_cols]
rownames(count_matrix) <- ensembl_ids
rm(ensembl_ids, keep_cols)
```


#Mutation Analysis

##1 Visualize mutation summaries
Find patients in mutation data that exists in all data sets.
```{r}
temp_mut_data <- mutation_data |>
  mutate(Tumor_Sample_Barcode = substr(Tumor_Sample_Barcode, 1, 12))

patient_data <- patient_data |>
  mutate(Tumor_Sample_Barcode = PATIENT_ID)

laml = read.maf(maf = temp_mut_data, clinicalData = patient_data)
rm(temp_mut_data)
```

```{r}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
```

```{r}
#oncoplot for top twenty mutated genes.
oncoplot(maf = laml, top = 20)
```

## Filter Silent and Missense and Regenerate MAF Summary
```{r}
mutation_filtered <- subset(mutation_intersect, Variant_Classification != "Silent")
mutation_filtered <- subset(mutation_filtered, Variant_Classification != "Missense_Mutation")
mutation_filtered <- subset(mutation_filtered,IMPACT =="HIGH")

count_matrix_filtered <- count_matrix_intersect[,unique(mutation_filtered$PATIENT_ID)]
rownames(count_matrix_filtered) <- count_matrix_gene
```


```{r, results='hide'}
laml.fil = read.maf(maf = mutation_filtered, clinicalData = patient_intersect)
write.mafSummary(maf = laml.fil, basename = 'laml_fil')
```


```{r}
plotmafSummary(maf = laml.fil, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
oncoplot(maf = laml, top = 20)
```


```{r}
hugo <- as.data.frame(table(mutation_filtered$Hugo_Symbol))
```

```{r}
##Create the onco matrix
cnv_events = unique(mutation_filtered$Variant_Classification)
oncomat = reshape2::dcast(
  data = mutation_filtered,
  formula = Hugo_Symbol ~ PATIENT_ID,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
```


```{r}
rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
oncomat.ordered <- oncomat[order(-hugo$Freq),]
mat <- oncomat.ordered
mat[mat!=""]=1
mat[mat==""]=0
mat <- apply(mat, 2 ,as.numeric)
mat <- as.matrix(mat)
rownames(mat)  <-  row.names(oncomat.ordered)
```

```{r}
reduce.mat <- mat[1:10,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE,
         clustering_method = "ward.D")
```
```{r}
cluster <- as.data.frame(cutree(res$tree_col, k = 2)) 

```


## Classify Patients
```{r}
library(dbplyr)
hugo.ordered <- rownames(reduce.mat)

survdata <- patient_intersect |> filter(PATIENT_ID %in% rownames(cluster))

# survdata <- survdata[,c("PATIENT_ID","OS_MONTHS","DSS_MONTHS","OS_STATUS","DAYS_LAST_FOLLOWUP","PFS_STATUS")]


survdata$Group <- cluster[,1]
survdata$deceased <- grepl("1",survdata$PFS_STATUS,fixed=TRUE)
survdata$overall_survival <- ifelse(survdata$deceased,
                                    survdata$OS_MONTHS*30,
                                    survdata$DAYS_LAST_FOLLOWUP)
# survdata$overall_survival <- survdata$OS_MONTHS
```

## Survival Analysis
```{r}
library(survival)
library(TCGAbiolinks)
library(survminer)
library(SummarizedExperiment)
fit <- survfit(Surv(overall_survival,deceased)~Group,data=survdata)
ggsurvplot(fit,data=survdata,pval=T)
```



## Expression Analysis
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
countData <- count_matrix_filtered #just to keep the same variable name in the tutorial
colData <- as.data.frame(survdata$Group)
rownames(colData) <- survdata$PATIENT_ID
colnames(colData) <- c("condition")
sampleDists = dist(t(countData),upper = TRUE)


annot_col = data.frame(colData$condition)
row.names(annot_col) <- rownames(colData)

sampleDistMatrix = as.matrix( sampleDists )
rownames(sampleDistMatrix) = colnames(countData)
colnames(sampleDistMatrix) = colnames(countData)

# pheatmap(sampleDistMatrix,
#          clustering_distance_rows = sampleDists,
#          clustering_distance_cols = sampleDists,
#          cluster_rows=FALSE, show_rownames=TRUE,
#          cluster_cols=TRUE,
#          annotation_col=annot_col)
# 
# pca_res <- prcomp(tcountData), scale. = TRUE)
# score <- pca_res$x
# 
# score = as.data.frame(score)
# score$color <- as.factor(colData$condition)
# 
# 
# ggplot(score, aes(x=PC1, y=PC2,  color=color)) +
#   geom_point(size = 4)
```


```{r}
library(DESeq2)
dds = DESeqDataSetFromMatrix(countData=countData,
                              colData=colData,
                              design=~condition)
dds = DESeq(dds)

res1 <- results(dds)
```


```{r}
summary(res1)
table(res1$padj<0.05)
```
```{r}
res_temp <- res1
res_temp <- res_temp[order(res_temp$pvalue),]
head(res_temp,10)
```
## Get top 10 genes from each side
```{r}
#Segment plot into two df
res_df <- as.data.frame(res_temp) |> drop_na()
res_df.lower <- res_df |> filter(log2FoldChange < 0)
res_df.upper <- res_df |> filter(log2FoldChange >= 0)

res_df.lower <- res_df.lower |> 
  mutate(impact_score = abs(log2FoldChange) * -log(pvalue)) |>
  arrange(desc(impact_score)) |>
  mutate(top = ifelse(row_number() <= 20, "Yes", "No"))

res_df.upper <- res_df.upper |> 
  mutate(impact_score = abs(log2FoldChange) * -log(pvalue)) |>
  arrange(desc(impact_score)) |>
  mutate(top = ifelse(row_number() <= 20, "Yes", "No"))

res_df <- rbind(res_df.upper, res_df.lower)

plot(res_df$log2FoldChange, -log(res_df$pvalue), col = factor(res_df$top))

top_df <- res_df |> filter(top == "Yes")
top_df
```





## adding gene name data (code from r-bloggers)
```{r}
entrez.symbols = mapIds(org.Hs.eg.db,
                     keys=row.names(res_temp), 
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

fully.formed.genes <- entrez.symbols[which(!is.na(entrez.symbols))]
res_formed <- res_temp[names(fully.formed.genes), ]
rownames(res_formed) <- fully.formed.genes
head(res_formed, 10)
```

## pathway analysis (code adapted from r-bloggers)
```{r}
library(pathview)
library(gage)
library(gageData)
data(kegg.sets.hs)
data(sigmet.idx.hs)
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
```
```{r}
foldchanges = res_formed$log2FoldChange
names(foldchanges) = res_formed$entrez
head(foldchanges)
```
```{r}

keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)

lapply(keggres, head)
```
Pull top 5 pathways
```{r}
keggrespathways = data.frame(id=rownames(keggres$greater), keggres$greater) %>% 
  as_tibble() %>% 
  filter(row_number()<=5) %>% 
  .$id %>% 
  as.character()
keggrespathways
```
```{r}
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```
```{r}
# Define plotting function for applying later
plot_pathway = function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="hsa", new.signature=FALSE)

# plot multiple pathways (plots saved to disk and returns a throwaway list object)
tmp = sapply(keggresids, function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="hsa"))
```



## pca plot
```{r}
# pca_res <- prcomp(t(countData), scale. = TRUE)
# score <- pca_res$x
# 
# score = as.data.frame(score)
# score$color <- as.factor(colData$condition)
# 
# 
# ggplot(score, aes(x=PC1, y=PC2,  color=color)) + 
#   geom_point(size = 4)
```


