---
title: "Testing"
author: "Aden Chan"
date: "2023-12-08"
output: html_document
---

---
title: "Differential Expression Analysis"
author: "Group 39 Amy"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
if (!require("pacman")) install.packages("pacman")
pacman::p_load(BiocManager, tidyverse, reshape2, pheatmap)

BiocManager::install("DESeq2")
BiocManager::install("maftools")

library(DESeq2)
library(tidyverse)
library(maftools)
```

## Reading Data
```{r}
count_matrix <- read_csv(file = "RNAseq_PRAD.csv")
mutation_data <- read_tsv(file = "data_mutations.txt")
patient_data <- read_tsv(file = "data_clinical_patient.txt", skip = 4)
```

## Find Patients with Complete Data
Get Patient ID in Mutation Data
```{r}
mutation_data <- mutation_data |> mutate(PATIENT_ID = substr(Tumor_Sample_Barcode, 1, 12))
```

Get Patient ID's from RNA Data
```{r}
RNA_ColNames <- count_matrix |> colnames() |> substr(1, 12)
sub_intersect <- intersect(patient_data$PATIENT_ID, mutation_data$PATIENT_ID)
main_intersect <- intersect(sub_intersect, RNA_ColNames)
head(main_intersect)
rm(RNA_ColNames, sub_intersect)
```
```{r}
patient_data <- patient_data |>
  filter(PATIENT_ID %in% main_intersect);
mutation_data <- mutation_data |>
  filter(PATIENT_ID %in% main_intersect);

ensembl_ids <- count_matrix |> select(...1) |> pull()
keep_cols <- count_matrix |> colnames()
keep_cols <- which(substr(keep_cols, 1, 12) %in% main_intersect)
count_matrix <- count_matrix[, keep_cols]
rownames(count_matrix) <- ensembl_ids
rm(ensembl_ids, keep_cols)
```


#Mutation Analysis

##1 Visualize mutation summaries
Find patients in mutation data that exists in all data sets.
```{r}
temp_mut_data <- mutation_data |>
  mutate(Tumor_Sample_Barcode = substr(Tumor_Sample_Barcode, 1, 12))

patient_data <- patient_data |>
  mutate(Tumor_Sample_Barcode = PATIENT_ID) |>
  mutate(PFS_STATUS = substr(PFS_STATUS, 1, 1)) |>
  mutate(PFS_DAYS = ifelse(PFS_STATUS == 1,
                            PFS_MONTHS*30,
                            DAYS_LAST_FOLLOWUP))

laml = read.maf(maf = temp_mut_data, clinicalData = patient_data)
rm(temp_mut_data)
```

```{r}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
```

```{r}
#oncoplot for top twenty mutated genes.
oncoplot(maf = laml, top = 20)
```

```{r}
hugo <- as.data.frame(table(mutation_data$Hugo_Symbol))
hugo.ordered <- hugo[order(-hugo$Freq),]

cnv_events = unique(mutation_data$Variant_Classification)
oncomat = reshape2::dcast(
  data = mutation_data,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)


rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]

oncomat.ordered <- oncomat[order(-hugo$Freq),]

mat <- oncomat.ordered
mat[mat!=""]=1
mat[mat==""]=0
mat <- apply(mat, 2 ,as.numeric)
mat <- as.matrix(mat)
rownames(mat)  <-  row.names(oncomat.ordered)

```

```{r}
reduce.mat <- mat[1:20,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE,
         clustering_method = "ward.D")
```

```{r}
patients.clustered <- cbind(patient_data, cluster = cutree(res$tree_col, k = 2))  |>
  mutate(PFS_STATUS = as.numeric(PFS_STATUS))
head(patients.clustered)
rm(fit, hugo, hugo.ordered, oncomat, oncomat.ordered, reduce.mat, res, cnv_events)
```


## Survival Analysis
```{r}
library(survival)
library(TCGAbiolinks)
library(survminer)
library(SummarizedExperiment)
fit <- survfit(Surv(PFS_DAYS,PFS_STATUS)~cluster, data=patients.clustered)
ggsurvplot(fit,data=patients.clustered,pval=T)
```



## Expression Analysis
Preprocessing
```{r}
#Subset count_matrix with fully formed patients
indexes <- count_matrix |>
  colnames() |>
  substr(1, 12)

gene_names <- rownames(count_matrix)
indexes <- which(indexes %in% main_intersect)
count_matrix <- count_matrix[, indexes]
rownames(count_matrix) <- gene_names
rm(indexes, gene_names)
```

```{r}
#Generate coldata dataframe
coldata <- as.data.frame(colnames(count_matrix))
colnames(coldata) <- c("sample_id")
coldata <- coldata |>
  mutate(PATIENT_ID = substr(sample_id, 1, 12)) |>
  inner_join(patients.clustered |> select(PATIENT_ID, cluster), by = join_by(PATIENT_ID)) |>
  select(-PATIENT_ID) |>
  column_to_rownames(var = "sample_id") |>
  mutate(cluster = as.factor(cluster))


#Filter count data for low counts
count_matrix.filtered <- count_matrix
rownames(count_matrix.filtered) <- rownames(count_matrix)

names <- rownames(count_matrix)[rowSums(count_matrix)>1]
count_matrix.filtered <- count_matrix[rowSums(count_matrix)>1,]
rownames(count_matrix.filtered) <- names
rm(names)

```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
library(DESeq2)

dds = DESeqDataSetFromMatrix(countData=count_matrix.filtered,
                              colData=coldata,
                              design=~cluster)
dds = DESeq(dds)

dds.results <- results(dds)
```
```{r}
dds.results
```

```{r}
summary(dds.results)
table(dds.results$padj < 0.05)
```

```{r}
dds.ordered <- dds.results[order(dds.results$padj),]
head(dds.ordered,10)
```
#MA Plot
```{r}
plotMA(dds.ordered, ylim=c(-2,2))
```
##Plot log2FoldChange vs -log(pval)
```{r}
plot(dds.ordered$log2FoldChange, -log(dds.ordered$padj))
```


## adding gene name data (code from r-bloggers)
```{r}
dds.ordered$symbol = mapIds(org.Hs.eg.db,
                     keys=gsub("\\..*","",row.names(dds.ordered)),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

head(dds.ordered, n = 10)
```

## pathway analysis (code adapted from r-bloggers)
```{r}
library(pathview)
library(gage)
library(gageData)
data(kegg.sets.hs)
data(sigmet.idx.hs)
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
```

```{r}
foldchanges = dds.ordered$log2FoldChange
names(foldchanges) = dds.ordered$symbol
head(foldchanges)
```
```{r}

keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)

lapply(keggres, head)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04510")
```


##Confirming Clusters
```{r}
vsd <- vst(dds)
```

```{r}
genes <- order(dds.ordered$padj, decreasing = TRUE)[1:10]
genes <- append(genes, order(dds.ordered$padj, decreasing = F)[1:10])

annot_col = data.frame(coldata$cluster)
row.names(annot_col) <- rownames(coldata)

sampleMatrix <- assay(vsd)[genes,]

rownames(sampleMatrix) = rownames(genes)
colnames(sampleMatrix) = colnames(count_matrix)

pheatmap(sampleMatrix , cluster_rows=FALSE, show_rownames=TRUE, show_colnames=F,
         cluster_cols=TRUE, annotation_col=annot_col, clustering_method = "ward.D2")
```












